src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java

```diff
@@ -499,4 +531,45 @@ private Filter getNotOverlappingFilter(
 
         return f;
     }
+
+    private class PercentagesManager {
+
+        private double[] percentages;
+        List<Double> collapsedPercentages;
+        boolean collapsePercentages;
+
+        public PercentagesManager(double[] percentages) {
+            this.percentages = percentages;
+            this.collapsePercentages =
+                    percentages != null && !(DoubleStream.of(percentages).sum() > 100.0);
+            this.collapsedPercentages = new ArrayList<>();
+        }
+
+        public void collapsePercentages(List<Rule> rules, Rule current, int index) {
+            if (outputPercentages && collapsePercentages) {
+                int ruleIdx = rules.indexOf(current);
+                double toSumTo = percentages[index];
+                double toBeSummed = collapsedPercentages.get(ruleIdx);
+                collapsedPercentages.set(ruleIdx, toBeSummed + toSumTo);
+            }
+        }
+
+        public void addPercentages(int currentIndex) {
+            if (outputPercentages) collapsedPercentages.add(percentages[currentIndex]);
+        }
+
+        public void addPercentagesToLabel(List<Rule> rules) {
+            if (outputPercentages) {
+                for (int i = 0; i < rules.size(); i++) {
+                    Rule rule = rules.get(i);
+                    StringBuffer sb =
```

Comment:
Java builds a StringBuffer every time you concatenate strings, avoid building it explicitly unless you're working on a loop/conditional where multiple string concatenations are split among different commands. 
For this one using:
``String title = rule.getDescription().getTitle() + " (" + collapsedPercentanges.get(i) + "%)"``
is just as efficient, but more readable.

Reply:
fixed